package me.simplicitee.project.addons.ability.chi;

import com.projectkorra.projectkorra.Element;
import com.projectkorra.projectkorra.GeneralMethods;
import com.projectkorra.projectkorra.ability.AddonAbility;
import com.projectkorra.projectkorra.ability.ChiAbility;
import com.projectkorra.projectkorra.ability.CoreAbility;
import com.projectkorra.projectkorra.ability.StanceAbility;
import com.projectkorra.projectkorra.attribute.Attribute;
import com.projectkorra.projectkorra.chiblocking.WarriorStance;
import com.projectkorra.projectkorra.util.ActionBar;
import me.simplicitee.project.addons.ProjectAddons;
import me.simplicitee.project.addons.util.versionadapter.PotionEffectAdapter;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Location;
import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import java.util.ArrayList;
import java.util.List;

public class NinjaStance extends ChiAbility implements AddonAbility, StanceAbility {
	
	@Attribute(Attribute.DURATION)
	private long stealthDuration;
	@Attribute(Attribute.DURATION)
	private long duration;
	
	private boolean stealth, stealthReady, stealthStarted;
	private long stealthStart, stealthChargeTime, stealthReadyStart, stealthCooldown;
	private List<PotionEffect> effects = new ArrayList<>();
	private PotionEffect invis = new PotionEffect(PotionEffectType.INVISIBILITY, 5, 2, true, false);	

	public NinjaStance(Player player) {
		super(player);
		
		if (bPlayer.isOnCooldown(this)) {
			return;
		}

		final StanceAbility stance = this.bPlayer.getStance();
		if (stance instanceof AddonAbility) {
			((CoreAbility)stance).remove();
			if (stance instanceof NinjaStance) {
				this.bPlayer.setStance(null);
				return;
			}
		}

		PotionEffectAdapter effectAdapter = ProjectAddons.instance.getPotionEffectAdapter();

		duration = ProjectAddons.instance.getConfig().getLong("Abilities.Chi.NinjaStance.Duration");
		stealthDuration = ProjectAddons.instance.getConfig().getLong("Abilities.Chi.NinjaStance.Stealth.Duration");
		stealthChargeTime = ProjectAddons.instance.getConfig().getLong("Abilities.Chi.NinjaStance.Stealth.ChargeTime");
		stealthCooldown = ProjectAddons.instance.getConfig().getLong("Abilities.Chi.NinjaStance.Stealth.Cooldown");
		effects.add(new PotionEffect(PotionEffectType.SPEED, 5, ProjectAddons.instance.getConfig().getInt("Abilities.Chi.NinjaStance.SpeedAmplifier") - 1, true, false));
		effects.add(new PotionEffect(effectAdapter.getJumpBoostPotionEffectType(), 5, ProjectAddons.instance.getConfig().getInt("Abilities.Chi.NinjaStance.JumpAmplifier") - 1, true, false));
		this.bPlayer.setStance(this);
		start();
		//bPlayer.setStance(this);
		GeneralMethods.displayMovePreview(player);
		player.playSound(player.getLocation(), Sound.ENTITY_ENDER_DRAGON_HURT, 0.2F, 2F);
	}

	@Override
	public long getCooldown() {
		return ProjectAddons.instance.getConfig().getLong("Abilities.Chi.NinjaStance.Cooldown");
	}

	@Override
	public Location getLocation() {
		return player.getLocation();
	}

	@Override
	public String getName() {
		return "NinjaStance";
	}

	@Override
	public boolean isHarmlessAbility() {
		return true;
	}

	@Override
	public boolean isSneakAbility() {
		return true;
	}

	@Override
	public void progress() {
		if (!player.isOnline() || player.isDead()) {
			remove();
			return;
		}
		if (!this.bPlayer.canBendIgnoreBinds(this) || !this.bPlayer.hasElement(Element.CHI)) {
			remove();
			return;
		}

		if (duration != -1 && System.currentTimeMillis() > getStartTime() + duration) {
			remove();
			return;
		}

		if (stealth) {
			if (System.currentTimeMillis() >= stealthStart + stealthChargeTime) {
				stealthReady = true;
			}
			
			if (!stealthStarted) {
				if (stealthReady && !player.isSneaking()) {
					stealthReadyStart = System.currentTimeMillis();
					stealthStarted = true;
				} else if (!player.isSneaking()) {
					stopStealth();
					return;
				}
				
				GeneralMethods.displayColoredParticle(stealthReady && player.isSneaking() ? "00ee00" : "000000", player.getEyeLocation().add(player.getEyeLocation().getDirection()));
			} else {
				if (System.currentTimeMillis() >= stealthReadyStart + stealthDuration) {
					stopStealth();
					bPlayer.addCooldown("ninjastealth", stealthCooldown);
				} else {
					player.addPotionEffect(invis);
				}
			}
		}
		
		player.addPotionEffects(effects);
	}

	@Override
	public void remove() {
		super.remove();
		this.player.playSound(this.player.getLocation(), Sound.ENTITY_ENDER_DRAGON_SHOOT, 0.5F, 2F);
		bPlayer.setStance(null);
		bPlayer.addCooldown(this);
	}
	
	@Override
	public String getAuthor() {
		return "Simplicitee";
	}

	@Override
	public String getVersion() {
		return ProjectAddons.instance.version();
	}

	@Override
	public void load() {}

	@Override
	public void stop() {}
	
	@Override
	public String getDescription() {
		return "This stance allows chiblockers to become faster and more stealthy (like a ninja)!";
	}
	
	@Override
	public String getInstructions() {
		return "Left click to begin to this stance > Hold sneak to begin stealth mode";
	}
	
	@Override
	public boolean isEnabled() {
		return ProjectAddons.instance.getConfig().getBoolean("Abilities.Chi.NinjaStance.Enabled");
	}

	public void beginStealth() {
		if (stealth) {
			ActionBar.sendActionBar(ChatColor.RED + "!> already cloaked <!", player);
			return;
		} else if (bPlayer.isOnCooldown("ninjastealth")) {
			ActionBar.sendActionBar(ChatColor.RED + "!> cooldown <!", player);
			return;
		}
		stealth = true;
		stealthStart = System.currentTimeMillis();
	}
	
	public void stopStealth() {
		stealth = false;
		stealthReady = false;
		stealthStarted = false;
	}
	
	public boolean isStealthed() {
		return stealth && stealthStarted;
	}
	
	public static double getDamageModifier() {
		return ProjectAddons.instance.getConfig().getDouble("Abilities.Chi.NinjaStance.DamageModifier");
	}

	@Override
	public String getStanceName() {
		return "NinjaStance";
	}
}
